rewrite-core/src/main/java/org/openrewrite/GitRemote.java

```diff
@@ -0,0 +1,178 @@
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite;
+
+import lombok.Value;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.jgit.transport.URIish;
+import org.openrewrite.marker.GitProvenance;
+
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Value
+public class GitRemote {
+    Service service;
+    String url;
+    String origin;
+    String path;
+
+    @Nullable
+    String organization;
+
+    String repositoryName;
+
+    public enum Service {
+        GitHub,
+        GitLab,
+        Bitbucket,
+        BitbucketCloud,
+        AzureDevOps,
+        Unknown
+    }
+
+    public static class Parser {
+        private final Map<String, Service> origins;
+
+        public Parser() {
+            origins = new LinkedHashMap<>();
+            origins.put("github.com", Service.GitHub);
+            origins.put("gitlab.com", Service.GitLab);
+            origins.put("bitbucket.org", Service.BitbucketCloud);
+            origins.put("dev.azure.com", Service.AzureDevOps);
+            origins.put("ssh.dev.azure.com", Service.AzureDevOps);
+        }
+
+        public Parser registerRemote(Service service, String origin) {
+            if (origin.startsWith("https://") || origin.startsWith("http://") || origin.startsWith("ssh://")) {
+                origin = new Parser.HostAndPath(origin).concat();
+            }
+            if (origin.contains("@")) {
+                origin = new Parser.HostAndPath("https://" + origin).concat();
+            }
+            if (service == Service.Unknown) {
+                // Do not override a known with an unknown service
+                origins.putIfAbsent(origin, service);
+            } else {
+                origins.put(origin, service);
+            }
+            return this;
+        }
+
+        public GitRemote parse(String url) {
+            Parser.HostAndPath hostAndPath = new Parser.HostAndPath(url);
+
+            String origin = hostAndPath.host;
+            Service service = origins.get(origin);
+            if (service == null) {
+                for (String maybeOrigin : origins.keySet()) {
+                    if (hostAndPath.concat().startsWith(maybeOrigin)) {
+                        service = origins.get(maybeOrigin);
+                        origin = maybeOrigin;
+                        break;
+                    }
+                }
+            }
+
+            if (service == null) {
+                // If we cannot find a service, we assume the last 2 path segments are the organization and repository name
```

Comment:
I don't think this will work with on-prem gitlab repos. You can name the host anything and have an arbitrary number of nested subgroups.

Reply:
correct, so for an on-prem gitlab we do require to register the base url